### PR TITLE
chore(ioredis): merge callback and promise signatures

### LINKED ISSUE #23637: Ioredis is not compatible with CompatibleRedisClient[]
I tried using the `@types/redlock` package and had problem:
 ```Argument of type 'Redis' is not assignable to parameter of type 'CompatibleRedisClient[]'. Property 'push' is missing in type 'Redis'.```
I'm using ioredis 3.1.4 as my Redis package. 
cc @andy-ms @BendingBender any ideas how to solve this? Thank you.

### LINKED ISSUE #27391: @types/ioredis accesing nodes() on type Redis
- I tried using the `@types/ioredis` package and had problems when issuing commands without id.
- Authors: @plantain-00 @chrisleck @aliarham11 @br8h @shaharmor @palindrom615

code example folowing guidelines on ioredis
https://github.com/luin/ioredis/issues/336

```
import { Cluster } from 'ioredis';

const cluster = new Cluster([
    { host: '127.0.0.1', port: 7000 },
    { host: '127.0.0.1', port: 7001 },
    { host: '127.0.0.1', port: 7002 },
]);

 cluster.on('ready', () => {
     console.log(cluster.nodes().length);
     console.log(cluster.nodes('slave').length);
});
```
compilation issue :
TS2339: property 'nodes' doesn't exist on types 'Redis'.

looking at d.ts file, interface Cluster#new returns a type Redis but Redis doesn't expose nodes().

workaround
The above code works perfectly if i delete @types/ioredis from package.json however i lose typing :)

### LINKED ISSUE #27784: [@types/ioredis] Several commands have return type "any"
- [x] I tried using the `@types/ioredis` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @plantain-00 @chrisleck @aliarham11 @br8h @shaharmor 
<hr>

Hi, several commands exposed by ioredis have a return type of `any` (not even `Promise<any>`). Examples of these commands are `mget`, `hgetall`, and `smembers`, but there are many more.

Seems like Redis responses are uniform enough to be able to type every potential response, and the type definitions should at least be making it clear that the methods are returning promises. In the case of `mget` and a few others, there isn't a separate overload for callbacks and promises either.

Is there a technical reason for all this or was it just that the authors didn't have time to fully flesh out the entire API?

[edit] Furthermore it seems commands that have return types defined may have them defined incorrectly, e.g. `get` is defined as returning `Promise<string>` when it seems like it should be `Promise<string|undefined>` for missing keys.

### LINKED ISSUE #28220: [@types/ioredis] psubscribe interface doesn't match api
**Expected:** psubscribe is supposed to accept a callback function as specified by [API](https://github.com/luin/ioredis#pubsub).

**Actual:** accepts array of strings ([link](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ioredis/index.d.ts#L422))
`psubscribe(...patterns: string[]): any;`
 

- [x] I tried using the `@types/ioredis` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @plantain-00 @pietrovismara @jeffkenney @br8h @andy-ms @minestarks @aliarham11 @BastienAr @d-yokoi 

If you do not mention the authors the issue will be ignored.

### LINKED ISSUE #34662: [@type/ioredis] blpop/brpop don't have timeout/callback parameter
If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/ioredis` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @plantain-00 @chrisleck @aliarham11 @br8h @shaharmor @palindrom615 @reconbot @funthing @iamolegga @tingwai-to @pettyalex

### LINKED ISSUE #35566: types of ioredis missing sendCommand and unlink
I tried using the `@types/ioredis` package and had problems.

### LINKED ISSUE #40324: [@types/ioredis] Missing commands
- [x] I tried using the `@types/ioredis` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @plantain-00 @chrisleck @aliarham11 @br8h @shaharmor

In `ioredis`, `zrangebylex()`, `zremrangebylex()` and `zrevrangebylex()` exist hovewer they don't exist in `@types/ioredis`.

Please create a pull request to fix it.

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/luin/ioredis
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

This is still somewhat WIP.

I'm trying to fix various issues while retaining type safety.
Unfortunately many of those changes result in breaking changes due to TS limitations.

Fixes #40324
Fixes #35566
FIxes #34662
Fixes #28220
Fixes #40930
Fixes #27784 (technically fixed by previous PR but hey :) )
Fixes #27391 (Not fixed by me but this but it's easier to bulk close
Fixes #23637 (This was also fixed a while ago but can be closed)

### CODE REVIEW

**Comment by SimonSchick** [comment]
Hey @kyleamazza I just created the common interface for the commands I mentioned in your issue, can you give those typings a spin?

**Comment by kyleamazza** [comment]
working for me 👍

**Comment by kyleamazza** [comment]
@SimonSchick Do you know what the status of this PR is? If this one has breaking changes, I can recreate the Commands interface in #40930 to avoid breaking changes.

**Comment by SimonSchick** [comment]
Waiting for approvals/reviews.

**Comment by pc-jedi** [comment]
@SimonSchick What is the status of this PR? I would like to see it merged soon :-)

**Comment by SimonSchick** [comment]
See previous comment 🤷‍♂

File: types/ioredis/ioredis-tests.ts [hunk 1]
```diff
@@ -10,15 +10,15 @@ redis.get('foo', (err, result) => {
...
 // Static check that returned value is always a number
-redis.del('foo', 'bar').then(result => result * 1);
+redis.del(['foo', 'bar']).then(result => result * 1);
 
 // Or using a promise if the last argument isn't a function
 redis.get('foo').then((result: string | null) => {
     console.log(result);
 });
 
 // Arguments to commands are flattened, so the following are the same:
```
Comment (pc-jedi) [inline types/ioredis/ioredis-tests.ts]: They are not the same anymore.

File: types/ioredis/index.d.ts [hunk 2]
```diff
@@ -678,192 +594,198 @@ declare namespace IORedis {
...
-        psubscribe(...patterns: string[]): Pipeline;
+        psubscribe(patterns: string | string[], callback?: Callback<number>): Promise<number>;
 
-        punsubscribe(...patterns: string[]): Pipeline;
+        punsubscribe(patterns: string | string[], callback?: Callback<number>): Promise<number>;
 
-        publish(channel: string, message: string, callback?: (err: Error, res: number) => void): Pipeline;
+        publish(channel: string, message: string, callback?: Callback<number>): Pipeline;
 
         watch(...keys: KeyType[]): Pipeline;
```
Comment (pc-jedi) [inline types/ioredis/index.d.ts]: What about `watch` of the `Pipeline` interface? This function does also have a callback, right?
Same for other methods of the `Pipeline` interface.

File: types/ioredis/ioredis-tests.ts [hunk 3]
```diff
@@ -329,7 +329,6 @@ cluster.defineCommand('defineCommand', {
     numberOfKeys: 1,
     lua: 'lua'
 });
-cluster.sendCommand();
```
Comment (pc-jedi) [inline types/ioredis/ioredis-tests.ts]: Shouldn't it be `send_command`?

**Review by pc-jedi — COMMENTED** [review]
I feel we have to add some more tests. But maybe it is easier to have a new issue/PR for that.

**Pull request closed without merging**